Add discussion "Distribution package vs. import package" (#1426)
pypa/packaging.python.org: source/discussions/distribution-package-vs-import-package.rst

the import package ``PIL``. This is for historical reasons: Pillow
started as a fork of the PIL library, thus it kept the import name
``PIL`` so that existing PIL users could switch to Pillow with little
effort. More generally, a fork of an existing library is a common reason
for differing names between the distribution package and the import
package.

On a given package index (like PyPI), distribution package names must be
unique. On the other hand, import packages have no such requirement.
Import packages with the same name can be provided by several
distribution packages. Again, forks are a common reason for this.

Conversely, a distribution package can provide several import packages,
although this is less common. An example is the attrs_ distribution
package, which provides both an ``attrs`` import package with a newer
API, and an ``attr`` import package with an older but supported API.

How do distribution package names and import package names compare?
===================================================================

Import packages should have valid Python identifiers as their name (the
:ref:`exact rules <python:identifiers>` are found in the Python

> documentation). In particular, they use underscores ``_`` as word

Author: ```suggestion
documentation) #[non-identifier-mod-name]_. In particular, they use underscores ``_`` as word
```

separator and they are case-sensitive.

On the other hand, distribution packages can use hyphens ``-`` or
underscores ``_``. They can also contain dots ``.``, which is sometimes
used for packaging a subpackage of a :ref:`namespace package
<packaging-namespace-packages>`. For most purposes, they are insensitive
to case and to ``-`` vs.  ``_`` differences, e.g., ``pip install
Awesome_Package`` is the same as ``pip install awesome-package`` (the
precise rules are given in the :ref:`name normalization specification
<name-normalization>`).

> 

Author: ```suggestion

-
---------------------------

.. [#non-identifier-mod-name] Although is technically possible
   to import packages/modules that do not have a valid Python identifier as
   their name, using :doc:`importlib <python:library/importlib>`,
   this is vanishingly rare and strongly discouraged.
```

although this is less common. An example is the attrs_ distribution
package, which provides both an ``attrs`` import package with a newer
API, and an ``attr`` import package with an older but supported API.

How do distribution package names and import package names compare?
===================================================================

Import packages should have valid Python identifiers as their name (the
:ref:`exact rules <python:identifiers>` are found in the Python
documentation) [#non-identifier-mod-name]_. In particular, they use underscores ``_`` as word
separator and they are case-sensitive.

On the other hand, distribution packages can use hyphens ``-`` or
underscores ``_``. They can also contain dots ``.``, which is sometimes
used for packaging a subpackage of a :ref:`namespace package
<packaging-namespace-packages>`. For most purposes, they are insensitive
to case and to ``-`` vs.  ``_`` differences, e.g., ``pip install
Awesome_Package`` is the same as ``pip install awesome-package`` (the
precise rules are given in the :ref:`name normalization specification
<name-normalization>`).

---------------------------

> .. [#non-identifier-mod-name] Although is technically possible

Author: ```suggestion
.. [#non-identifier-mod-name] Although it is technically possible
```

the import package ``PIL``. This is for historical reasons: Pillow
started as a fork of the PIL library, thus it kept the import name
``PIL`` so that existing PIL users could switch to Pillow with little
effort. More generally, a fork of an existing library is a common reason
for differing names between the distribution package and the import
package.

On a given package index (like PyPI), distribution package names must be
unique. On the other hand, import packages have no such requirement.
Import packages with the same name can be provided by several
distribution packages. Again, forks are a common reason for this.

Conversely, a distribution package can provide several import packages,
although this is less common. An example is the attrs_ distribution
package, which provides both an ``attrs`` import package with a newer
API, and an ``attr`` import package with an older but supported API.

How do distribution package names and import package names compare?
===================================================================

Import packages should have valid Python identifiers as their name (the
:ref:`exact rules <python:identifiers>` are found in the Python

> documentation) #[non-identifier-mod-name]_. In particular, they use underscores ``_`` as word

Author: ```suggestion
documentation) [#non-identifier-mod-name]_. In particular, they use underscores ``_`` as word
```

separator and they are case-sensitive.

On the other hand, distribution packages can use hyphens ``-`` or
underscores ``_``. They can also contain dots ``.``, which is sometimes
used for packaging a subpackage of a :ref:`namespace package
<packaging-namespace-packages>`. For most purposes, they are insensitive
to case and to ``-`` vs.  ``_`` differences, e.g., ``pip install
Awesome_Package`` is the same as ``pip install awesome-package`` (the
precise rules are given in the :ref:`name normalization specification
<name-normalization>`).

> -

Author: ```suggestion

```

processing. Its distribution package name is ``Pillow``, but it provides
the import package ``PIL``. This is for historical reasons: Pillow
started as a fork of the PIL library, thus it kept the import name
``PIL`` so that existing PIL users could switch to Pillow with little
effort. More generally, a fork of an existing library is a common reason
for differing names between the distribution package and the import
package.

On a given package index (like PyPI), distribution package names must be
unique. On the other hand, import packages have no such requirement.
Import packages with the same name can be provided by several
distribution packages. Again, forks are a common reason for this.

Conversely, a distribution package can provide several import packages,
although this is less common.

How do distribution package names and import package names compare?
===================================================================

Import packages should have valid Python identifiers as their name.  In
particular, they use underscores ``_`` as word separator and they are
case-sensitive.

On the other hand, distribution packages can use hyphens ``-`` or

> underscores ``.``. They can also contain dots ``.``, which is sometimes

Reviewer: ```suggestion
underscores ``_``. They can also contain dots ``.``, which is sometimes
```

used for packaging a subpackage of a :ref:`namespace package
<packaging-namespace-packages>`. For most purposes, they are insensitive
to case and to ``-`` vs.  ``_`` differences, e.g., ``pip install

> Awesome_Package`` is the same as ``pip install awesome-package``.

Reviewer: Maybe this could use a link to [Package name normalization](https://python-packaging-user-guide--1426.org.readthedocs.build/en/1426/specifications/name-normalization/)?

.. _distribution-package-vs-import-package:

=======================================
Distribution package vs. import package
=======================================

A number of different concepts are commonly referred to by the word
"package". This page clarifies the differences between two distinct but

> related meanings in Python packaging, "distribution package" and "import

Reviewer 2: @jeanas I must've overlooked that, thanks! Though, maybe we could drop "distribution" from that wording. Linux folks just say "package".

package".

What's a distribution package?
==============================

> A distribution package is a piece of software that you can install.

Reviewer 2: @willingc what's our style guide stance on using “you” in discussions? Should they be more formal (as in passive voice and “we”?)
I don't have a strong opinion, it's just something that caught my eye.

Reviewer 3: Good question @webknjaz. Right now, @jeanas is correct about what the style has been in the past.

I'm hoping to adopt a style guide (perhaps one that exists already like Google or RedHat) and use [vale]( https://vale.sh/) in CI to provide prose linting. But this will be an effort in 2024 likely in the PyCon timeframe.

Author: I'm going to resolve this thread, given that the PR is consistent with the style guide (it even gives the example “Right: You can install it by running…”).

Most of the time, this is synonymous with "project". When you type ``pip
install pkg``, or when you write ``dependencies = ["pkg"]`` in your
``pyproject.toml``, ``pkg`` is the name of a distribution package. When
you search or browse PyPI_, the most widely known centralized source for
installing Python software, what you see is a list of distribution
packages. Alternatively, the term "distribution package" can be used to
refer to a specific file that contains a certain version of a project.

Note that in the Linux world, "distribution package" refers to a package
provided by the system package manager, which is a different meaning.

What's an import package?
=========================

An import package is a Python module. Thus, when you write ``import
pkg`` or ``from pkg import func`` in your Python code, ``pkg`` is the
name of an import package. More precisely, import packages are special
Python modules that can contain submodules. For example, the ``numpy``
package contains modules like ``numpy.linalg`` and
``numpy.fft``. Usually, an import package is a directory on the file
system, containing modules as ``.py`` files and subpackages as

> subdirectories.

Reviewer 2: Alright, this can be polished later. Marking the thread as resolved.

You can use an import package as soon as you have installed a distribution
package that provides it.

What are the links between distribution packages and import packages?
=====================================================================

> By convention, a distribution package usually provides one single import

Reviewer 2: I'm not sure this is a known convention. My feeling is that it's by accident, mostly. Or by example…

Reviewer 2: I think, in the past, people also experimented with having “python2-” and “python3-” pairs.

Author: I went for "most of the time".

package (or non-package module), with a matching name. For example,
``pip install numpy`` lets you ``import numpy``.

However, this is only a convention. PyPI and other package indices do
not enforce any relationship between the name of a distribution package
and the import packages it provides.

A distribution package could provide an import package with a different
name. An example of this is the popular Pillow_ library for image
processing. Its distribution package name is ``Pillow``, but it provides
the import package ``PIL``. This is for historical reasons: Pillow
started as a fork of the PIL library, thus it kept the import name
``PIL`` so that existing PIL users could switch to Pillow with little
effort. More generally, a fork of an existing library is a common reason
for differing names between the distribution package and the import
package.

On a given package index (like PyPI), distribution package names must be
unique. On the other hand, import packages have no such requirement.

> Import packages with the same name can be provided by several

Reviewer 4: I'm +0 here: I don't think it would *hurt* to have this here, but I see the point about keeping the document focused.

This is probably expressed in equivalent language below, but I think the security concern can be distilled down to: "there is no relationship between a PyPI package name and an `import`-able package name. If you assume such a relationship, you may end up installing unintended, potentially malicious, packages."

> Import packages with the same name can be provided by several

Author: Thanks, I've added a sentence that boils down to this.

distribution packages. Again, forks are a common reason for this.

Conversely, a distribution package can provide several import packages,

> although this is less common.

Reviewer 2: Perfect! Resolving...

> How do distribution package names and import package names compare?

Reviewer 2: > My humble opinion: either none of them, or the first one in a footnote mentioning that it is technically possible to import modules with names that aren't valid Python identifiers, but this is vanishingly rare and not recommended.

Yeah, that sounds reasonable.

> How do distribution package names and import package names compare?

Author: Done.

===================================================================

> Import packages should have valid Python identifiers as their name.  In

Reviewer 2: This is a good place to link the CPython docs:
```suggestion
Import packages should have valid :ref:`Python identifiers <python:identifiers>` as their name.  In
```

.. _distribution-package-vs-import-package:

=======================================
Distribution package vs. import package
=======================================

A number of different concepts are commonly referred to by the word
"package". This page clarifies the differences between two distinct but
related meanings in Python packaging, "distribution package" and "import
package".

What's a distribution package?
==============================

A distribution package is a piece of software that you can install.
Most of the time, this is synonymous with "project". When you type ``pip
install pkg``, or when you write ``dependencies = ["pkg"]`` in your
``pyproject.toml``, ``pkg`` is the name of a distribution package. When

> you search or browse PyPI_, the most widely known centralized source for

Reviewer 2: ```suggestion
you search or browse the PyPI_, the most widely known centralized source for
```

> installing Python software, what you see is a list of distribution

Reviewer 2: Is "software" too generic? It's mostly libs+frameworks.

Author: ```suggestion
installing Python libraries and tools, what you see is a list of distribution
```

Author: OK. I've used “Python libraries and tools” to accounts for CLI tools.

packages. Alternatively, the term "distribution package" can be used to
refer to a specific file that contains a certain version of a project.

> Note that in the Linux world, "distribution package" refers to a package

Reviewer 2: This needs clarification. "Distro" is what a GNU/Linux flavor is called, like Fedora, or Arch, or Gentoo, or RHEL. In that context, that's the only "distribution" people talk about. They call the software bundles "packages", just "packages".

Can we at least drop the word "distribution"?

Could also link something like https://linuxsimply.com/linux-basics/package-management/what-is-package/ / https://en.wikipedia.org/wiki/Linux_distribution.

Author: ```suggestion
Note that in the Linux world, a "distribution package",
most commonly abbreviated as "distro package" or just "package",
```

> provided by the system package manager, which is a different meaning.

Author: ```suggestion
is something provided by the system package manager of the `Linux distribution <distro_>`_,
which is a different meaning.
```

Conversely, a distribution package can provide several import packages,
although this is less common. An example is the attrs_ distribution
package, which provides both an ``attrs`` import package with a newer
API, and an ``attr`` import package with an older but supported API.

How do distribution package names and import package names compare?
===================================================================

Import packages should have valid Python identifiers as their name (the
:ref:`exact rules <python:identifiers>` are found in the Python
documentation). In particular, they use underscores ``_`` as word
separator and they are case-sensitive.

On the other hand, distribution packages can use hyphens ``-`` or
underscores ``_``. They can also contain dots ``.``, which is sometimes
used for packaging a subpackage of a :ref:`namespace package
<packaging-namespace-packages>`. For most purposes, they are insensitive
to case and to ``-`` vs.  ``_`` differences, e.g., ``pip install
Awesome_Package`` is the same as ``pip install awesome-package`` (the
precise rules are given in the :ref:`name normalization specification
<name-normalization>`).

> 

Author: ```suggestion

.. _distro: https://en.wikipedia.org/wiki/Linux_distribution
```
